### PR TITLE
MGMT-18876: Update admission controller to block changes to ICI once the installation started

### PR DESCRIPTION
**Pull request by eranco74**
This change also removes the handleing for BMH update post install since the effect of such update is unknown
Note that the webhook doesn't handle changes to resources referenced by the ICI

### CODE REVIEW

File: controllers/imageclusterinstall_controller.go [hunk 1]
```diff
@@ -227,13 +226,6 @@ func (r *ImageClusterInstallReconciler) Reconcile(ctx context.Context, req ctrl.
 		}
 	}
 
-	if ici.Status.BareMetalHostRef != nil && !v1alpha1.BMHRefsMatch(ici.Spec.BareMetalHostRef, ici.Status.BareMetalHostRef) {
```
Comment (tsorya) [inline controllers/imageclusterinstall_controller.go:230]: why? I am not sure we need it too much but still, the idea is that you can sort of change bmh and will be nice to cleanup the old one but we can drop it i believe

**Comment by tsorya** [comment]
lets split to 2 prs maybe?

File: api/v1alpha1/imageclusterinstall_webhook.go [hunk 2]
```diff
@@ -53,21 +56,6 @@ func (r *ImageClusterInstall) ValidateCreate() (admission.Warnings, error) {
 	return nil, nil
 }
 
-func (r *ImageClusterInstall) validate() error {
```
Comment (tsorya) [inline api/v1alpha1/imageclusterinstall_webhook.go:56]: Should not it run  on create too?

Comment (tsorya) [reply api/v1alpha1/imageclusterinstall_webhook.go:56]: sorry, seems like it just changed it place

File: api/v1alpha1/imageclusterinstall_webhook.go [hunk 3]
```diff
@@ -81,20 +69,43 @@ func (r *ImageClusterInstall) ValidateUpdate(old runtime.Object) (admission.Warn
...
+	if err := isValidMachineNetwork(r.Spec.MachineNetwork); err != nil {
+		return fmt.Errorf("invalid machine network: %w", err)
+	}
+	return nil
+}
+
+func installationStarted(r *ImageClusterInstall) bool {
+	// Block update if the image is ready
+	// Consider updating this to check the ClusterInstallCompleted condition
+	imageReadyCondition := findCondition(r.Status.Conditions, hivev1.ClusterInstallRequirementsMet)
```
Comment (tsorya) [inline api/v1alpha1/imageclusterinstall_webhook.go:105]: hmm, i am not sure we update this condition anymore to say the truth, I think better to validate if there is bmh in status or even in reference

File: api/v1alpha1/imageclusterinstall_webhook.go [hunk 4]
```diff
@@ -81,20 +69,43 @@ func (r *ImageClusterInstall) ValidateUpdate(old runtime.Object) (admission.Warn
...
 	return nil, nil
 }
 
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *ImageClusterInstall) ValidateDelete() (admission.Warnings, error) {
+	return nil, nil
+}
+
+func (r *ImageClusterInstall) validate() error {
+	if err := isValidSSHPublicKey(r.Spec.SSHKey); err != nil {
```
Comment (tsorya) [inline api/v1alpha1/imageclusterinstall_webhook.go:89]: Maybe add validation that we have referenced bmh?

File: api/v1alpha1/imageclusterinstall_webhook.go (hunk 3 above)
Comment (tsorya) [reply api/v1alpha1/imageclusterinstall_webhook.go:105]: Or we can add this condition here https://github.com/openshift/image-based-install-operator/blob/main/controllers/imageclusterinstall_controller.go#L277

File: api/v1alpha1/imageclusterinstall_webhook.go (hunk 4 above)
Comment (eranco74) [reply api/v1alpha1/imageclusterinstall_webhook.go:89]: What for?

Comment (tsorya) [reply api/v1alpha1/imageclusterinstall_webhook.go:89]: as we don't support ici without bmh anymore? but it's ok if we allow to set it after

File: api/v1alpha1/imageclusterinstall_webhook.go [hunk 5]
```diff
@@ -142,19 +153,11 @@ func isValidMachineNetwork(machineNetwork string) error {
...
-}
-
-func BMHRefsMatch(ref1 *BareMetalHostReference, ref2 *BareMetalHostReference) bool {
-	if ref1 == nil && ref2 == nil {
-		return true
-	}
-
-	if ref1 == nil && ref2 != nil || ref1 != nil && ref2 == nil {
-		return false
+func findCondition(conditions []hivev1.ClusterInstallCondition, condType hivev1.ClusterInstallConditionType) *hivev1.ClusterInstallCondition {
```
Comment (tsorya) [inline api/v1alpha1/imageclusterinstall_webhook.go:156]: can't you use function from conditions?

File: api/v1alpha1/imageclusterinstall_webhook_test.go [hunk 6]
```diff
@@ -240,43 +273,24 @@ var _ = Describe("ValidateUpdate", func() {
...
+			Message: InstallSucceededMessage,
 		}}
 
 		warns, err := newClusterInstall.ValidateUpdate(oldClusterInstall)
 		Expect(warns).To(BeNil())
 		Expect(err).To(BeNil())
 	})
 
-	It("metadata update succeeds when BMH ref is set", func() {
+	It("metadata update succeeds when iamge is ready", func() {
```
Comment (tsorya) [inline api/v1alpha1/imageclusterinstall_webhook_test.go:293]: nit: iamge -> image

**Pull request merged**